cpp/src/arrow/device.h

```diff
@@ -29,6 +29,25 @@
 
 namespace arrow {
 
+/// \brief EXPERIMENTAL: Device type enum which matches up with C Data Device types
+enum class DeviceType : char {
+  UNKNOWN = 0,
+  CPU = 1,
+  CUDA = 2,
```

Comment:
Should we consider adding a seperate allocation type for allocations that are made using Stream Ordered Allocators [More Details Here](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html) These allocators are asynchronous with respect to the host. When a user allocates data using a stream ordered allocator the allocator returns immediately and the allocation is guaranteed to be valid before the next operation is added to that particular stream. Because arrow is not handling cuda streams if a user uses the asynchronous allocator then there is no guarantee this allocation will be performed before we attempt to write or read to that memory. 

This means that if we are using async allocators the user should be informed this is the case so that they can do things like capture the stream from whatever library is making the allocation to handle the stream or at least call cudaDeviceSynchronize() to ensure that this allocation is indeed already available.

Reply:
It's intended that knowing which devices are stream ordered is up to the users of the library. This is why the DeviceArray includes an sync event that can be populated and if it isn't null it needs to be synchronized on.

Additionally after this PR, I'll be working on an interface to abstract more of the event and synchronization for buffers. I don't think we need a separate allocation type for async allocators, we just need to ensure that we're consistent with the device type enums.

Reply:
Its not a device which is stream ordered but rather an allocation. A user could have allocations that are both made via async allocators and via synchronous ones. If the memory layout doesn't have a way of specifying this then users will have to basically have an extra piece of information which is carried around with every allocation they make in order to ensure  that asynchronously allocated memory  calls synchronize before being accessed. 

The more I think about it though there is a further complication which is knowing which stream it was allocated on and the fact that if you enable per thread default stream this would further complicate situations in which every thread would have its own stream of value 0.  I can understand why it might be necessary to hold off on this for a while but I do think this is something that should be kept in  mind.

Reply:
I agree

Reply:
I agree

Reply:
@felipeblazing That makes sense, I'll add that to my list of things for this project. For now I'd prefer to hold off on adding those complexities to a future PR rather than this one. Unless you think that this PR in and of itself is going to be problematic without it?

Reply:
This is a bit of an edge case. If only one process was performing allocations and it was consistent with its allocation strategy, which most are, then you would not really have this issue. So yes its fine to hold off!